### DISCUSSION TITLE
What library is used in `poetry` for pretty printing errors?

### DISCUSSION THREAD

**Discussion by abitrolly** [python-poetry/poetry#3499] [Q&A] [answered]
^^^

**Comment by sinoroc** [comment]
* https://github.com/python-poetry/poetry/pull/2230
* https://pypi.org/project/cleo/
* https://pypi.org/project/pastel/
* https://pypi.org/project/crashtest/ ~~https://pypi.org/project/woops/~~

All from the author or _poetry_.

**Reply by abitrolly** [reply]
Looks like the error handling and formatting is done by `whoops`, but there are no docs or repo for it https://github.com/sdispater/clikit/issues/23

**Comment by sdispater** [comment, marked as answer]
To add to what @sinoroc said:

- `woops` has been replaced by `crashtest` (https://github.com/sdispater/crashtest, same library different name) which is responsible for manipulating and inspecting Python exceptions in an easier way. It also provides a "solution provider" system so that you can provide solutions for know exceptions.
- The actual error formatting is done in `clikit` (https://github.com/sdispater/clikit/blob/master/src/clikit/ui/components/exception_trace.py) – which willl soon be integrated back into `cleo` – by using `crashtest` internally.
- Poetry uses `cleo` at its heart for its CLI which, in turn, uses `clikit`.

**Reply by abitrolly** [reply]
Is there any chance to get pretty exception handler+printer available as a standalone library with stable API? I would sugar up a few open source projects that throw exceptions way too often to be friendly for new users.
